### DISCUSSION TITLE
Opening a URL on a browser for my Node.JS app

### DISCUSSION THREAD

**Discussion by remoteconn-7891** [community/community#77238] [Other Feature Feedback, Questions, & Ideas]
### Select Topic Area

Product Feedback

### Body

So I'm trying to connect the frontend via Flutter (Client) to the backend via Express.JS (Server) through REST API. However when I type in 'http://localhost:3000' I get a connection error. I will send a screenshot of the error page and provide the code for both the index.js (backend for login screen) and main.dart (frontend login) and I can confirm that my server is indeed listening on PORT 3000 and it is connected to MongoDB database. I will however omitt the session secret code and the MongoDB usenname-password for security purposes.  I'm considering using EC2 to set up an IP address to help with this issue. Please any kind of help

index.js

`const bcrypt = require("bcrypt");
const express = require("express");
const cors = require("cors");
const session = require('express-session');
const passport = require("passport");
const LocalStrategy = require('passport-local').Strategy;
const mongoose = require("mongoose");
const passportLocalMongoose = require("passport-local").Strategy;
const User = require("./model/user");
const authenticateJWT = require('./model/middleware/JWTMiddleware');
const path = require('path');

const app = express();

// Make sure CORS is properly set 
app.use(cors());
// EJS for view engine
app.set("view engine", "ejs");

// View directory set up
app.set("views", path.join(__dirname,"views"));

// middleware function for express-session
app.use(
    session({
        secret: '',
        resave: false,
        saveUninitialized: false,
    })
);  

app.use(express.urlencoded({ extended: true }));

mongoose.connect("")
.then(() => {
    console.log("MongoDB connection successful:");
})
.catch(err => {
    console.error("MongoDB connection error:", err);
});

app.use(passport.initialize());
app.use(passport.session());

passport.use(new LocalStrategy({
    usernameField: 'email',
    passwordField: 'password'
},
function (username, password, done) {
    // user authentication here
}
));

// Login page
app.get("/login", function(req, res, next) {
    res.render("login");
    next();
});

// Login function for returning user
app.post("/login", passport.authenticate("local", {
    successRedirect: "/profile",
    failureRedirect: "/login",
    failureFlash: true
}));

// Route for profile screens
app.get("/profile", (req, res, next) => {
    const { role } = req.user;
    next()

    if (role === 'job_seeker') {
        res.redirect("/job_seeker_profile");
    } else if (role === 'company') {
        res.redirect("/company_profile");
    } else {
        res.status(403).json({ message: 'Unauthorized access' });
    }
});

        // Profile page for Job Seeker
        app.get("/job_seeker_profile", (req, res, next) => {
            res.render("job_seeker_profile");
            next()
    });
    
        // Profile page for company
        app.get("/company_profile", (req, res, next) => {
            res.render("company_profile");
            next()
        });

// Route for user create an accout
app.post("/signup", async (req, res) => {
const {email, password } = req.body;
    try {
        // Make sure the password is hashed before storing it in MongoDB
        const saltRounds = 10;
        const hashedPassword = await bcrypt.hash(password, saltRounds);

        // Register new user account with hashed password
        const user = await User.create({
            email,
            password: hashedPassword,
            role: req.body.role || 'remote job_seeker'
        });
       

        // Message sent for login error
        return res.status(200).json(user);       
    } catch (error) {
        console.error("Error registering user:", error);
        return res.status(500).json({ error: 'Internal server error' });
    }

//JWT middleware function
app.post("/check-token", authenticateJWT, (req, res, next) => {
    res.json({ message: 'Valid token' });
    next()
});

// Route that compares entered plaintext password with hashed password
app.post("/check-password", async (req, res, next) => {
       // Assumed plaintext password
    const plaintextPassword = req.body.password;

    // Fetch the hashed password
const hashedPassword = "hashed password";

try {
    const isPasswordValid = await bcrypt.compare(plaintextPassword, hashedPassword);

    if (isPasswordValid) {
        res.json({ message: 'Valid password' });
    } else {
        res.status(401).json({ message: 'Invalid password' });
    }
}catch (error) {
    console.error("Error verifying password:", error);
    res.status(500).json({ error: 'Internal server error' });
}
});

//server is listening on PORT 3000
app.listen(3000, '0.0.0.0', () => {
    console.log("Server is listening on PORT 3000");
})});`

main.dart

`import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

void main() {
  runApp(const MainApp());
}

class MainApp extends StatelessWidget {
  const MainApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Virtualify',
      theme: ThemeData(
        primarySwatch: Colors.red,
      ),
      home: Login(title: 'Welcome to Virtualify'),
    );
  }
}

class Login extends StatefulWidget {
  const Login({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<Login> createState() => LoginState();
}

class LoginState extends State<Login> {
  int _counter = 0;
  final _formKey = GlobalKey<FormState>(); // Move _formKey here

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  TextEditingController emailController = TextEditingController();
  TextEditingController passwordController = TextEditingController();

  Future<void> _login() async {
    final response = await http.post(
      Uri.parse('http://localhost:3000'),
      headers: <String, String>{
        'Content-Type': 'application/json; charset=UTF-8',
      },
      body: jsonEncode(<String, String>{
        'email': emailController.text,
        'password': passwordController.text
      }),
    );

    if (response.statusCode == 200) {
      // Successful login, response was handled
      print('Successful login');
    } else {
      // Login unsuccessful
      print('Unsuccessful login');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'This button was pushed x amount of times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
            Form(
              key: _formKey, // Use _formKey here
              child: Padding(
                padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 18),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 18),
                      child: TextFormField(
                        controller: emailController,
                        decoration: const InputDecoration(
                          border: OutlineInputBorder(),
                          labelText: "Email",
                        ),
                        validator: (value) {
                          if (value == null || value.isEmpty) {
                            return 'Enter a valid email';
                          }
                          return null;
                        },
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 18),
                      child: TextFormField(
                        controller: passwordController,
                        obscureText: true,
                        decoration: const InputDecoration(
                          border: OutlineInputBorder(),
                          labelText: "Password",
                        ),
                        validator: (value) {
                          if (value == null || value.isEmpty) {
                            return 'Enter a valid password';
                          }
                          return null;
                        },
                      ),
                    ),
                    Padding(
                      padding: const EdgeInsets.symmetric(horizontal: 10, vertical: 18.0),
                      child: Center(
                        child: ElevatedButton(
                          onPressed: () {
                            if (_formKey.currentState!.validate()) {
                              _login();
                              // User will navigate to the home screen
                            } else {
                              ScaffoldMessenger.of(context).showSnackBar(
                                const SnackBar(content: Text('Enter all fields')),
                              );
                            }
                          },
                          child: const Text('Submit'),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

class DropdownMenu extends StatefulWidget {
  const DropdownMenu({Key? key}) : super(key: key);

  @override
  _DropdownMenuState createState() => _DropdownMenuState();
}

class _DropdownMenuState extends State<DropdownMenu> {
  String selectRole = 'job seeker'; // Default user role

  @override
  Widget build(BuildContext context) {
    return DropdownButton<String>(
      value: selectRole,
      onChanged: (String? newValue) {
        setState(() {
          selectRole = newValue!;
        });
      },
      items: <String>['job seeker', 'company']
          .map<DropdownMenuItem<String>>((String value) {
        return DropdownMenuItem<String>(
          value: value,
          child: Text(value),
        );
      }).toList(),
    );
  }
}`

**Comment by remoteconn-7891** [comment]
<img width="1440" alt="Screenshot 2023-11-28 at 12 05 23 PM" src="https://github.com/community/community/assets/142646120/c64eb35b-11ff-4712-a0ce-d8500536f2b0">

**Comment by remoteconn-7891** [comment]
And ChatGPT is practically useless at this point, I tried that route already

**Comment by remoteconn-7891** [comment]
PS I also some more files related to the node.JS backend and I'm also trying to configure my Godaddy domain I purchased a few months to my Flutter app through DNS. So please let me know if you need anything else from me

**Comment by leeringrose** [comment]
@remoteconn-7891 
Can you please share your code inside the code block for more understandability?
Have you been sure if the backend node engine is running without any errors output on the console.

**Reply by remoteconn-7891** [reply]
What code block, inside index.js? The first code I pasted for Node.JS is the all of index.js. Yes, it was running on PORT 3000 and was connected to MongoDB. PS I am getting errors on the console for the backend now, but that's only because I was using userMail to send email verification link to email use nodeMail

**Comment by remoteconn-7891** [comment]
I realize it's a shitload of code lines and I apologize for that. But here is the screenshot of current error on the terminal for more clarity.
<img width="1440" alt="Screenshot 2023-11-29 at 1 03 24 PM" src="https://github.com/community/community/assets/142646120/0f68b8b5-2d32-4284-8381-b8f62ae496e8">

**Reply by remoteconn-7891** [reply]
Here is the new index.js, I also commented out the transport & userMail block of code

`const bcrypt = require("bcrypt");
const express = require("express");
const cors = require("cors");
const session = require('express-session');
const passport = require("passport");
const mongoose = require("mongoose");
const LocalStrategy = require("passport-local").Strategy;
const passportLocalMongoose = require("passport-local-mongoose");
const User = require("./model/user");
const authenticateJWT = require('./model/middleware/JWTMiddleware');
const path = require('path');
const nodemailer = require("nodemailer");

const app = express();

// Send email verification link to email via nodemail and transport

const transporter = nodemailer.createTransport({
    service: 'outlook',
    auth: {
        user: 'youremail@outlook.com',
        pass: 'yourpassword'
    }
})

app.get('/verify-email', (req, res, next) => {
    const userMail = 'useremail.com';
    next();
})

const verificationLink = 'https://yourwebsite.com/verify?token=...'; // Replace with your actual verification link'

const mailOptions = {
    from: 'youremail@outlook.com',
    to: userMail,
    subject: 'Email verification',
    text: 'Verify email by clicking on this link ${verificationLink}'
};

transporter.sendMail(mailOptions, (error, info) => {
    if (error) {
        return console.error('Error sending email:', error);
    }
    console.log('Email successfully sent:', info.response);
    res.send('Email sent for varification');
});


// Make sure CORS is properly set 
app.use(cors());
// EJS for view engine
app.set("view engine", "ejs");

// View directory set up
app.set("views", path.join(__dirname,"views"));

// middleware function for express-session
app.use(
    session({
        secret: '$^%FGH*#%347dbh',
        resave: false,
        saveUninitialized: false,
    })
);  

app.use(express.urlencoded({ extended: true }));

mongoose.connect("mongodb+srv://coreyjames689:zXzreMzalyYOVgNa@cluster0.jybj21u.mongodb.net/?retryWrites=true&w=majority")
.then(() => {
    console.log("MongoDB connection successful:");
})
.catch(err => {
    console.error("MongoDB connection error:", err);
});

app.use(passport.initialize());
app.use(passport.session());

passport.use(new LocalStrategy({
    usernameField: 'email',
    passwordField: 'password'
},
function (username, password, done) {
    // user authentication here
}
));

// Login page
app.get("/login", function(req, res, next) {
    res.render("login");
    next();
});

// Login function for returning user
app.post("/login", passport.authenticate("local", {
    successRedirect: "/profile",
    failureRedirect: "/login",
    failureFlash: true
}));

// Route for profile screens
app.get("/profile", (req, res, next) => {
    const { role } = req.user;
    next()

    if (role === 'job_seeker') {
        res.redirect("/job_seeker_profile");
    } else if (role === 'company') {
        res.redirect("/company_profile");
    } else {
        res.status(403).json({ message: 'Unauthorized access' });
    }
});

        // Profile page for Job Seeker
        app.get("/job_seeker_profile", (req, res, next) => {
            res.render("job_seeker_profile");
            next()
    });
    
        // Profile page for company
        app.get("/company_profile", (req, res, next) => {
            res.render("company_profile");
            next()
        });

// Route for user create an accout
app.post("/signup", async (req, res) => {
const {email, password } = req.body;
    try {
        // Make sure the password is hashed before storing it in MongoDB
        const saltRounds = 10;
        const hashedPassword = await bcrypt.hash(password, saltRounds);

        // Register new user account with hashed password
        const user = await User.create({
            email,
            password: hashedPassword,
            role: req.body.role || 'remote job_seeker'
        });
       

        // Message sent for login error
        return res.status(200).json(user);       
    } catch (error) {
        console.error("Error registering user:", error);
        return res.status(500).json({ error: 'Internal server error' });
    }

//JWT middleware function
app.post("/check-token", authenticateJWT, (req, res, next) => {
    res.json({ message: 'Valid token' });
    next()
});

// Route that compares entered plaintext password with hashed password
app.post("/check-password", async (req, res, next) => {
       // Assumed plaintext password
    const plaintextPassword = req.body.password;

    // Fetch the hashed password
const hashedPassword = "hashed password";

try {
    const isPasswordValid = await bcrypt.compare(plaintextPassword, hashedPassword);

    if (isPasswordValid) {
        res.json({ message: 'Valid password' });
    } else {
        res.status(401).json({ message: 'Invalid password' });
    }
}catch (error) {
    console.error("Error verifying password:", error);
    res.status(500).json({ error: 'Internal server error' });
}
});

//server is listening on PORT 3000
app.listen(3000, '0.0.0.0', () => {
    console.log("Server is listening on PORT 3000");
})});
`

**Comment by remoteconn-7891** [comment]
Other than the code I just sent you now, if you could specify exactly where inside you code you need to see other than the transport/nodeMail code.

**Reply by leeringrose** [reply]
I was talking about the code you put here. Just put it with ensuring the whole code is in the code block that this platform supports.

I put the example below for your understanding.

You could start with a triple of "`" and end also with those for your whole code part.

```
const app = () => {
  return (<div>
<h1>Code inside code panel `feature</h1>
</div>
}
```

**Reply by remoteconn-7891** [reply]
What I meant to ask was, was there a specific location of my code you wanted to see? For example, the signup or login route bodies of code or the block of code containing PassPort. I'm a little confused here, I apologize. But what is this, it almost looks like a little like HTML when the backend is running on Node.JS, specifically Express.JS. Also I was able to fix the errors in my terminal since I took out some of the code for the userMail so it's connected to the database now, but for some reason it's not listening on PORT 3000.

**Comment by remoteconn-7891** [comment]
<img width="1440" alt="Screenshot 2023-11-29 at 6 08 06 PM" src="https://github.com/community/community/assets/142646120/d0ca2547-f31f-4bf1-b47a-b081e9f938eb">

**Comment by remoteconn-7891** [comment]
Ironically I can open up a login page via Xcode iOS simulator. Still have problems here. But I'm trying to test this app outside of Xcode on another device. Which is one of the reasons I chose the use Node.JS as a backend.
<img width="1440" alt="Screenshot 2023-11-30 at 8 37 50 AM" src="https://github.com/community/community/assets/142646120/8449b1da-6af4-48e7-b7b9-c7ccd7b2c64d">

**Comment by remoteconn-7891** [comment]
Are you sure you know what you're talking? My project is running on express.JS on the backend. The code you provided me is clearly HTML, this should be node. Also I viewed your profile, it reads as frontend development

**Reply by remoteconn-7891** [reply]
I don't think you know what you're talking about. I just Googled that code you sent me, it relates to React. Thanks anyway
